tests/keras/engine/test_training.py

```diff
@@ -808,5 +808,50 @@ def test_target_tensors():
                          sample_weight={'dense_a': np.random.random((10,))})
 
 
+@keras_test
+def test_model_custom_target_tensors():
+    a = Input(shape=(3,), name='input_a')
+    b = Input(shape=(3,), name='input_b')
+
+    a_2 = Dense(4, name='dense_1')(a)
+    dp = Dropout(0.5, name='dropout')
+    b_2 = dp(b)
+
+    y = K.placeholder([10, 4], name='y')
+    y1 = K.placeholder([10, 3], name='y1')
+    y2 = K.placeholder([7, 5], name='y2')
+    model = Model([a, b], [a_2, b_2])
+
+    optimizer = 'rmsprop'
+    loss = 'mse'
+    loss_weights = [1., 0.5]
+
+    # test list of target tensors
+    with pytest.raises(ValueError):
+        model.compile(optimizer, loss, metrics=[], loss_weights=loss_weights,
+                      sample_weight_mode=None, target_tensors=[y, y1, y2])
+    model.compile(optimizer, loss, metrics=[], loss_weights=loss_weights,
+                  sample_weight_mode=None, target_tensors=[y, y1])
+    input_a_np = np.random.random((10, 3))
+    input_b_np = np.random.random((10, 3))
+
+    output_a_np = np.random.random((10, 4))
+    output_b_np = np.random.random((10, 3))
+
+    out = model.train_on_batch([input_a_np, input_b_np],
+                               [output_a_np, output_b_np], {y: np.random.random((10, 4)), y1: np.random.random((10, 3))})
```

Comment:
@fchollet I see what you're saying, it is a difference between two valid APIs. 

I have a specific use case with reinforcement learning algorithms, where input tensors can be implemented to load large input data from a simulation without passing through python, but iterative calculations are much smaller and can run conveniently in python. Here is some quick pseudocode I wrote on the fly based on my real use case:

```python

image_tensor_in = Input(tensor=image_input_tensor)
action_tensor_in = Input()
cnn = my_cnn(image_tensor_in, action_tensor_in)
cnn.compile(..., input_tensor=[image_tensor_in, action_tensor_in])

# first iteration continuous action space, random actions
actions = np.array(np.random.random(...))
# an input tensor is used to move the large data
action_scores = Model.train_on_batch(...[None, actions]...)

# resample with Cross Entropy Method for better possible actions
resampled_actions = CEM(actions, action_scores)
action_scores = Model.train_on_batch(...[None, resampled_actions]...)

```

Those inputs are augmented by on the fly per-batch processing in python with algorithms such as the Cross Entropy Method to iteratively optimize a large parameter space of inputs to the network. While implementing a python tensor is certainly possible, from my perspective simply passing in a numpy array to train is much cleaner, easier to use, more convenient, and consistent with the existing API. In my case CEM only works with small arrays of floating point numbers output by another part of the algorithm that can be sped up with a library like [numba](https://numba.pydata.org/), but the overall network has large images that benefit from input tensors' high performance.

Essentially, I saw placeholders as simply another data tensor we must account for, one that happens to need a numpy input to be fed in later. Accounting for this is why `is_placeholder`, `_make_ins`, and the changes to `backend.Function` are made over in `training.py` and `topology.py` in #6928 .